leapp/reporting/__init__.py

```diff
@@ -5,102 +5,256 @@
+    """Report title"""
+    @property
+    def path(self):
+        return ('title',)
+
+
+class Summary(BasePrimitive):
+    """Report summary"""
+    @property
+    def path(self):
+        return ('summary',)
+
+
+class Severity(BasePrimitive):
+    """Report severity"""
+
+    INFO = 'info'
+    LOW = 'low'
+    MEDIUM = 'medium'
+    HIGH = 'high'
+
+    def __init__(self, value=None):
+        if not isinstance(value, str):
+            raise TypeError('Value of "{}" must be a string'.format(self.__class__.__name__))
+        severities = ('info', 'low', 'medium', 'high')
+        if value not in severities:
+            raise ValueError('Severity should be one of: "{}"'.format(' '.join(severities)))
+        self._value = value
+
+    @property
+    def path(self):
+        return ('severity',)
+
+
+class Audience(BasePrimitive):
+    """Target audience of the report"""
+    def __init__(self, value=None):
+        if not isinstance(value, str):
+            raise TypeError('Value of "{}" must be a string'.format(self.__class__.__name__))
+        audiences = ('sysadmin', 'developer')
+        if value not in audiences:
+            raise ValueError('Audience should be one of: "{}"'.format(' '.join(audiences)))
+        self._value = value
+
+    @property
+    def path(self):
+        return ('audience',)
+
+
+class Flags(BasePrimitive):
+    """Report flags"""
 
+    INHIBITOR = 'inhibitor'
+    FAILURE = 'failure'
 
-def report(title=None, detail=None, renderers=None, audience=None, flags=None, severity=None):
+    def __init__(self, value=None):
+        if not isinstance(value, list):
+            raise TypeError('Value of "{}" must be a list'.format(self.__class__.__name__))
+        self._value = value
+
+    @property
+    def path(self):
+        return ('flags',)
+
+
+class Tags(BasePrimitive):
+    """Report tags"""
+
+    ACCESSIBILITY = 'accessibility'
+    AUTHENTICATION = 'authentication'
+    COMMUNICATION = 'communication'
+    DRIVERS = 'drivers'
+    EMAIL = 'email'
+    FILESYSTEMS = 'filesystems'
+    HIGH_AVAILABILITY = 'high availability'
+    NETWORK = 'network'
+    SANITY = 'sanity'
+    SECURITY = 'security'
+    SELINUX = 'selinux'
+    SERVICES = 'services'
+    TIME_MANAGEMENT = 'time management'
+    TOOLS = 'tools'
+    UPGRADE_PROCESS = 'upgrade process'
+
+    def __init__(self, value=None):
+        if not isinstance(value, list):
+            raise TypeError('Value of "{}" must be a list'.format(self.__class__.__name__))
+        self._value = value
+
+    @property
+    def path(self):
+        return ('tags',)
+
+
+class ExternalLink(BasePrimitive):
+    """External link report detail field"""
+    def __init__(self, url=None, title=None):
+        if not all(isinstance(v, str) for v in (url, title)):
+            raise TypeError('Values "url" and "title" of "{}" must be a string'.format(self.__class__.__name__))
+        self._value = {'url': url, 'title': title}
+
+    @property
```

Comment:
Open question: introduce class BaseListPrimitive?

Reply:
@shaded-enmity what do you think. Shall wee need any extension soon? I am not against but probably in the next iteration...

Reply:
Imho overriding is_list method in all list-based classes that follow isn't as clear as introducing another BaseList type and using inheritance where necessary.

Reply:
@fernflower you are free to submit a PR to this PR :)

Reply:
But why have 2 if this code isn't merged yet? :)

Reply:
Well you make a PR to his branch on his fork, he merges in his branch then. It's not 2 PRs to this repo.

Reply:
applied...